docs/examples/backend.py

```diff
@@ -0,0 +1,333 @@
+# applying the Softplus bijector to the parameter's value and projecting its value onto
+# the real line. Were the parameter's value to be closer to 0, then the transformation
+# would be more pronounced.
+
+# %%
+transform(PositiveReal(value=1e-6), DEFAULT_BIJECTION, inverse=True)
+
+# %% [markdown]
+# ### Transforming Multiple Parameters
+#
+# In the above, we transformed a single parameter. However, in practice your parameters
+# may be nested within several functions e.g., a kernel function within a GP model.
+# Fortunately, transforming several parameters is a simple operation that we here
+# demonstrate for a conjugate GP posterior (see our [Regression
+# Notebook](https://docs.jaxgaussianprocesses.com/examples/regression/) for detailed
+# explanation of this model.).
+
+# %%
+kernel = gpx.kernels.Matern32()
+meanf = gpx.mean_functions.Constant()
+
+prior = gpx.gps.Prior(mean_function=meanf, kernel=kernel)
+
+likelihood = gpx.likelihoods.Gaussian(100)
+posterior = likelihood * prior
+posterior
+
+# %% [markdown]
+# Now contained within the posterior PyGraph here there are four parameters: the
+# kernel's lengthscale and variance, the noise variance of the likelihood, and the
+# constant of the mean function. Using NNX, we may realise these parameters through the
+# `nnx.split` function. The `split` function deomposes a PyGraph into a `GraphDef` and
+# `State` object. As the name suggests, `State` contains information on the parameters'
+# state, whilst `GraphDef` contains the information required to reconstruct a PyGraph
+# from a give `State`.
+
+# %%
+from flax import nnx
+
+graphdef, state = nnx.split(posterior)
+state
+
+# %% [markdown]
+# The `State` object behaves just like a PyTree and, consequently, we may use JAX's
+# `tree_map` function to alter the values of the `State`. The updated `State` can then
+# be used to reconstruct our posterior. In the below, we simply increment each
+# parameter's value by 1.
+
+# %%
+import jax.tree_util as jtu
+
+updated_state = jtu.tree_map(lambda x: x + 1, state)
+updated_state
+
+# %% [markdown]
+# Let us now use NNX's `merge` function to reconstruct the posterior distribution using
+# the updated state.
+
+# %%
+updated_posterior = nnx.merge(graphdef, updated_state)
+updated_posterior
+
+# %% [markdown]
+# However, we begun this point of conversation with bijectors in mind, so let us now see
+# how bijectors may be applied to a collection of parameters in GPJax. Fortunately, this
+# is very straightforward, and we may simply use the `trasnform` function as before.
+
+# %%
+transformed_state = transform(state, DEFAULT_BIJECTION, inverse=True)
+transformed_state
+
+# %% [markdown]
+# We may also (re-)constrain the parameters' values by setting the `inverse` argument of
+# `transform` to False.
+
+# %%
+retransformed_state = transform(transformed_state, DEFAULT_BIJECTION, inverse=False)
+retransformed_state == transformed_state
```

Comment:
`transform` is pure, no?

Reply:
`transform` alters the value of the first argument, so I don't think it's pure. In this case, it alters the `transformed_state` variable.

Reply:
But it alters the value within the scope of the function. The function returns the same result if invoked with the same inputs and it has no side-effects outside of its scope, so it is pure by definition. 

Otherwise 
```python
def func(a, b):
    a += 1
    c = a + b
    return c
```
would also be impure.

Reply:
Francesco and I synced up offline and concluded that `transform` is indeed impure in its current state.